hazelcast/src/main/java/com/hazelcast/jet/impl/pipeline/transform/DistinctTransform.java

```diff
@@ -53,7 +56,34 @@ public void addToDag(Planner p, Context context) {
 
     @SuppressWarnings("unchecked")
     private static <T, K> ProcessorSupplier distinctP(FunctionEx<? super T, ? extends K> keyFn) {
-        return filterUsingServiceP(nonSharedService(pctx -> new HashSet<>()),
-                (seenItems, item) -> seenItems.add(keyFn.apply((T) item)));
+        return filterUsingServiceP(
+                nonSharedService(context -> new DistinctChecker<T, K>(keyFn, context.maxProcessorAccumulatedRecords())),
+                (checker, item) -> checker.isDistinct((T) item)
+        );
+    }
+
+    private static final class DistinctChecker<T, K> {
+
+        private final FunctionEx<? super T, ? extends K> keyFn;
+        private final long maxItems;
+        private final Set<K> seenItems;
+
+        private DistinctChecker(FunctionEx<? super T, ? extends K> keyFn, long maxItems) {
+            this.keyFn = keyFn;
+            this.maxItems = maxItems;
+            this.seenItems = new HashSet<>();
+        }
+
+        private boolean isDistinct(T item) {
+            if (seenItems.add(keyFn.apply(item))) {
+                if (seenItems.size() > maxItems) {
```

Comment:
This is some off-by-one trickery, in the other cases yesterday you used `==`. But I see you have the same test as for all the others :thinking:

Reply:
Passes:
```
Pipeline pipeline = Pipeline.create();
        pipeline.readFrom(TestSources.items(0, 1, 2, 3, 4, 4, 5))
                .groupingKey(wholeItem())
                .aggregate(counting())
                .writeTo(noop());

        assertThatThrownBy(() -> instance().newJob(pipeline).join())
                .hasMessageContaining("Exception thrown to prevent an OutOfMemoryError on this Hazelcast instance");
```
Fails:
```
    @Test
    public void when_maxAccumulatedRecordsCountIsExceededWhileGrouping_then_throws2() {
        Pipeline pipeline = Pipeline.create();
        pipeline.readFrom(TestSources.items(0, 1, 2, 3, 4, 4, 5))
                .groupingKey(wholeItem())
                .distinct()
                .writeTo(noop());

        assertThatThrownBy(() -> instance().newJob(pipeline).join())
                .hasMessageContaining("Exception thrown to prevent an OutOfMemoryError on this Hazelcast instance");
    }
```

Reply:
The thing is, in other cases I know that new item will be produced past the check, here I don't till I actually try to add it to the set.

Reply:
It depends whether you compare before or after adding. The issue here is that we actually accumulated one item more than we should. But if we checked the size before adding, then we would fail even if that new item wasn't distinct and wouldn't increase the size. So you can't do it perfectly...
I think it's ok here.